kvserver,allocator: decouple allocator from storepool

AlexTalks opened the pull request:
This change refactors the allocator to accept the store pool as an
argument, moving it closer to being stateless and purely functional.
This is done by moving to using the storepool through its primary 
owner (the store configuration), or through references held by 
users of the allocator.  The refactoring enables the the allocator to
consider potential scenarios rather than those simply based on the
current liveness.

Part of #91570.

Release Note: None

AlexTalks committed:
allocator: refactor allocator to accept StorePool arguments

This change adds methods to evaluate allocator actions and targets
utilizing a passed-in `StorePool` object, allowing for the allocator to
consider potential scenarios rather than those simply based on the
current liveness.

Depends on #91461, #91965.

Part of #91570.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91941)
<!-- Reviewable:end -->

kvoli reviewed (commented):
I'm not a fan of the decorated methods `...WithStorePool(...)`. It adds to the complexity of the allocator API by exposing additional methods.

Would it be possible to instead pass it as an argument for every top level function instead? 

This also relates to the approach taken on subsequent PRs - is it possible to use `planOneChange` in place of any forked methods.

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91941#-:-NJggNos12kDOKNbG75W:bjladal)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @andrewbaptist)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91941#-NJgbyvE2bWPdG6xp0YE):*
nit: drop the "to be"


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (commented):
This is doable - let me know if you think it makes more sense to _always_ pass in the `StorePool` on every allocator method (which may complicate things with callers), or if it would make sense to simply add an override parameter, like:
```
func (a *Allocator) ComputeAction(..., overrideStorePool storepool.AllocatorStorePool) {
  storePool := a.StorePool
  if overrideStorePool != nil {
     storePool = overrideStorePool
  }
  ...
}
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91941#-:-NJgmuwN18P8vwYJERrn:b-hbckgf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)



<!-- Sent from Reviewable.io -->

AlexTalks committed:
kvserver,asim: move storepool ownership out of allocator

This fully decouples the StorePool from the allocator by moving
ownership of the configured view of peer stores to the store
configuration (the original source), the store rebalancer, and various
structures in the allocator simulator.

Part of #91570.

Release note: None

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>   It would be good to add the storepool  to the replicate queue similar to the store rebalancer.

> Reviewed 1 of 9 files at r3, 23 of 23 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91941#-:-NJqj7P9As62J3eX5T6t:bpwi9wg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks and @andrewbaptist)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 614 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91941#-NJqfZ3O6gZu8RiiknD1:-NJqfZ3O6gZu8RiiknD2:b-3oej1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/79e7b0a84e9873ff3241bf774b4f0d2c5a38bf3f/pkg/kv/kvserver/replicate_queue.go#L614)):*
> ```Go
> ) (shouldQueue bool, priority float64) {
> 	desc, conf := repl.DescAndSpanConfig()
> 	action, priority := rq.allocator.ComputeAction(ctx, rq.store.cfg.StorePool, conf, desc)
> ```

I'm thinking it would be better if the replicate queue were instantiated with a ref to the StorePool similar to the store rebalancer.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 512 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/91941#-NJqgMpR0u95lKRMXPzG:-NJqgMpR0u95lKRMXPzH:b4q2ks9) ([raw file](https://github.com/cockroachdb/cockroach/blob/79e7b0a84e9873ff3241bf774b4f0d2c5a38bf3f/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L512)):*
> ```Go
> 
> // MakeAllocator creates a new allocator.
> // The deterministic flag indicates that this allocator is intended to be used
> ```

What if it is not? This seems pretty fragile. There is a deterministic test which should catch this however, so it should be okay if any unlucky soul runs into it.


<!-- Sent from Reviewable.io -->

AlexTalks commented:
bors r+
